### PR TITLE
Fix incompatibility of reflection with SQA 1.2.0

### PR DESCRIPTION
**Pull request by jklukas**
Addresses https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/issues/138

## Todos
- [x] MIT compatible
- [x] Tests
- [x] Documentation
- [x] Updated CHANGES.rst

### CODE REVIEW

**Comment by jklukas** [comment]
This gets reflection working on SQAlchemy 1.2.0+ while maintaining compatibility with older versions. It completely ignores the new comment functionality for now. Integrating that I leave to a separate change.

Tests are still failing on master due to a change in how key constraints are reflected by Redshift.

**Comment by jklukas** [comment]
cc @jleclanche

File: sqlalchemy_redshift/dialect.py [hunk 1]
```diff
@@ -604,6 +604,11 @@ def _get_table_or_view_names(self, relkind, connection, schema=None, **kw):
     def _get_column_info(self, *args, **kwargs):
         kw = kwargs.copy()
         encode = kw.pop('encode', None)
+        if sa.__version__ >= '1.2.0':
```
Comment (graingert) [inline sqlalchemy_redshift/dialect.py]: is there a distutils.version class we can use here?

Comment (graingert) [reply sqlalchemy_redshift/dialect.py]: How about a little bit of magic:

```python
try:
    from inspect import getfullargspec as getargspec
except ImportError:
    from inspect import getargspec

if 'comment' in getargspec(super(RedshiftDialect, self)._get_column_info).args:
    ...
```

Comment (graingert) [reply sqlalchemy_redshift/dialect.py]: Yo dawg I heard you like reflection

Comment (jklukas) [reply sqlalchemy_redshift/dialect.py]: Well that's fancy. Will try this.

File: sqlalchemy_redshift/dialect.py [hunk 2]
```diff
@@ -604,6 +609,12 @@ def _get_table_or_view_names(self, relkind, connection, schema=None, **kw):
     def _get_column_info(self, *args, **kwargs):
         kw = kwargs.copy()
         encode = kw.pop('encode', None)
+        if 'comment' in getargspec(
+                super(RedshiftDialect, self)._get_column_info).args:
```
Comment (graingert) [inline sqlalchemy_redshift/dialect.py]: could probably have a global:
```python
def args(fn):
    return getargspec(fn).args
```

to save a bit of space on this line.

File: sqlalchemy_redshift/dialect.py (hunk 1 above)
Comment (jleclanche) [reply sqlalchemy_redshift/dialect.py]: I strongly recommend against this. Reflection in Python is a nasty thing to introduce to production code, both for performance and predictability reasons. Checking `__version__` is a very fine way to achieve what you're doing.

**Comment by jleclanche** [comment]
lgtm but strong -1 on refection over version checking.

**Comment by jklukas** [comment]
I don't have previous experience with python reflection to understand the impacts here. The `sa.__version__ >= '1.2.0'` check is simpler and looks to be sufficient for our needs here, so I'm happy to revert to that.

@graingert - Do you think we gain something substantial from the reflection approach?

File: sqlalchemy_redshift/dialect.py [hunk 3]
```diff
@@ -604,6 +616,11 @@ def _get_table_or_view_names(self, relkind, connection, schema=None, **kw):
     def _get_column_info(self, *args, **kwargs):
         kw = kwargs.copy()
         encode = kw.pop('encode', None)
+        if 'comment' in arglist(super(RedshiftDialect, self)._get_column_info):
```
Comment (graingert) [inline sqlalchemy_redshift/dialect.py]: if we're worried about perf, how about a global cache of:

```python
has_comment = 'comment' in arglist(PGDialect_psycopg2._get_column_info)
```

Comment (jklukas) [reply sqlalchemy_redshift/dialect.py]: I'm leaning toward reverting to version check at this point. I don't see a clear benefit to doing the more complicated thing.

File: sqlalchemy_redshift/dialect.py (hunk 1 above)
Comment (jklukas) [reply sqlalchemy_redshift/dialect.py]: I've gone back to this simpler version check.

**Comment by jklukas** [comment]
@graingert - Are you okay with this PR as it now stands (with the SQA version check rather than introspection)?

If so, I'll merge this PR and cut a release so that more folks don't end up seeing breakages due to SQA version upgrade.

**Comment by jklukas** [comment]
Released 0.7.1: https://pypi.python.org/pypi?name=sqlalchemy-redshift&version=0.7.1&:action=display

**Pull request merged**